In [1]:
import numpy as np 
import pandas as pd
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
%matplotlib inline

In [5]:
df=pd.read_excel('data/多重共线性.xls')

In [6]:
df.head()

Y   X2    X3
0   70   80   810
1   65  100  1009
2   90  120  1273
3   95  140  1425
4  110  160  1633

Y 消费支出

X2 收入

X3 财富

In [7]:
df.describe()

Y          X2           X3
count   10.000000   10.000000    10.000000
mean   111.000000  170.000000  1740.000000
std     31.428932   60.553007   617.731153
min     65.000000   80.000000   810.000000
25%     91.250000  125.000000  1311.000000
50%    112.500000  170.000000  1754.500000
75%    135.000000  215.000000  2163.750000
max    155.000000  260.000000  2686.000000

In [8]:
results = smf.ols('Y ~ X2 +X3', data=df).fit()

In [10]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.964
Model:                            OLS   Adj. R-squared:                  0.953
Method:                 Least Squares   F-statistic:                     92.40
Date:                Tue, 03 Apr 2018   Prob (F-statistic):           9.29e-06
Time:                        16:22:53   Log-Likelihood:                -31.587
No. Observations:                  10   AIC:                             69.17
Df Residuals:                       7   BIC:                             70.08
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     24.7747      6.752      3.669      0.0

/Applications/anaconda/lib/python3.6/site-packages/scipy/stats/stats.py:1334: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


$\hat{Y}_i=24.7747 +0.9415X_{2i}-0.0424X_{3i}$   财富回归系数为负值，越有钱消费越低，与预期不一致

se= (6.752 )  (0.823 )  (0.081)

t=  (3.669 )  (1.144 )  (-0.526)  

p=  (0.008  )  (0.290 )  (0.615 )   回归系数均不显著 ，说明变量无影响

$调整R^2=0.953$  调整$R^2$非常高，说明模型拟合非常好

所有这些看起来怪异的现象，是由于收入与财富的高度相关

In [11]:
results = smf.ols('Y ~ X2', data=df).fit()

In [12]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.962
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     202.9
Date:                Tue, 03 Apr 2018   Prob (F-statistic):           5.75e-07
Time:                        19:21:50   Log-Likelihood:                -31.781
No. Observations:                  10   AIC:                             67.56
Df Residuals:                       8   BIC:                             68.17
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     24.4545      6.414      3.813      0.0

/Applications/anaconda/lib/python3.6/site-packages/scipy/stats/stats.py:1334: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


In [13]:
results = smf.ols('Y ~ X3', data=df).fit()

In [14]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.951
Method:                 Least Squares   F-statistic:                     176.7
Date:                Tue, 03 Apr 2018   Prob (F-statistic):           9.80e-07
Time:                        19:22:04   Log-Likelihood:                -32.444
No. Observations:                  10   AIC:                             68.89
Df Residuals:                       8   BIC:                             69.49
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     24.4110      6.874      3.551      0.0

/Applications/anaconda/lib/python3.6/site-packages/scipy/stats/stats.py:1334: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


In [15]:
results = smf.ols('X2 ~ X3', data=df).fit()

In [16]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     X2   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.998
Method:                 Least Squares   F-statistic:                     3849.
Date:                Tue, 03 Apr 2018   Prob (F-statistic):           5.06e-12
Time:                        19:22:25   Log-Likelihood:                -23.807
No. Observations:                  10   AIC:                             51.61
Df Residuals:                       8   BIC:                             52.22
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.3863      2.898     -0.133      0.8

/Applications/anaconda/lib/python3.6/site-packages/scipy/stats/stats.py:1334: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


$R^2非常高，表明两者高度相关$